File: pkg/code/code.go
// Package code defines Nexus bytecode opcodes and instruction encoding.
package code

import (
	"encoding/binary"
	"fmt"
)

// Instructions is a sequence of bytecode bytes.
type Instructions []byte

// Opcode is a single bytecode operation.
type Opcode byte

const (
	OpConstant Opcode = iota
	OpPop
	OpAdd
	OpSub
	OpMul
	OpDiv
	OpMod
	OpTrue
	OpFalse
	OpNull
	OpEqual
	OpNotEqual
	OpGreaterThan
	OpGreaterEqual
	OpLessThan
	OpLessEqual
	OpMinus
	OpBang
	OpJump
	OpJumpNotTruthy
	OpGetGlobal
	OpSetGlobal
	OpGetLocal
	OpSetLocal
	OpGetBuiltin
	OpArray
	OpHash
	OpIndex
	OpCall
	OpReturnValue
	OpReturn
	OpClosure
	OpGetFree
	OpCurrentClosure
	OpSetIndex
	OpMember
	OpAnd
	OpOr
)

// Definition describes an opcode's operands.
type Definition struct {
	Name          string
	OperandWidths []int
}

var definitions = map[Opcode]*Definition{
	OpConstant:       {"OpConstant", []int{2}},
	OpPop:            {"OpPop", []int{}},
	OpAdd:            {"OpAdd", []int{}},
	OpSub:            {"OpSub", []int{}},
	OpMul:            {"OpMul", []int{}},
	OpDiv:            {"OpDiv", []int{}},
	OpMod:            {"OpMod", []int{}},
	OpTrue:           {"OpTrue", []int{}},
	OpFalse:          {"OpFalse", []int{}},
	OpNull:           {"OpNull", []int{}},
	OpEqual:          {"OpEqual", []int{}},
	OpNotEqual:       {"OpNotEqual", []int{}},
	OpGreaterThan:    {"OpGreaterThan", []int{}},
	OpGreaterEqual:   {"OpGreaterEqual", []int{}},
	OpLessThan:       {"OpLessThan", []int{}},
	OpLessEqual:      {"OpLessEqual", []int{}},
	OpMinus:          {"OpMinus", []int{}},
	OpBang:           {"OpBang", []int{}},
	OpJump:           {"OpJump", []int{2}},
	OpJumpNotTruthy:  {"OpJumpNotTruthy", []int{2}},
	OpGetGlobal:      {"OpGetGlobal", []int{2}},
	OpSetGlobal:      {"OpSetGlobal", []int{2}},
	OpGetLocal:       {"OpGetLocal", []int{1}},
	OpSetLocal:       {"OpSetLocal", []int{1}},
	OpGetBuiltin:     {"OpGetBuiltin", []int{1}},
	OpArray:          {"OpArray", []int{2}},
	OpHash:           {"OpHash", []int{2}},
	OpIndex:          {"OpIndex", []int{}},
	OpCall:           {"OpCall", []int{1}},
	OpReturnValue:    {"OpReturnValue", []int{}},
	OpReturn:         {"OpReturn", []int{}},
	OpClosure:        {"OpClosure", []int{2, 1}},
	OpGetFree:        {"OpGetFree", []int{1}},
	OpCurrentClosure: {"OpCurrentClosure", []int{}},
	OpSetIndex:       {"OpSetIndex", []int{}},
	OpMember:         {"OpMember", []int{2}},
	OpAnd:            {"OpAnd", []int{}},
	OpOr:             {"OpOr", []int{}},
}

// Lookup returns the definition for op.
func Lookup(op Opcode) (*Definition, error) {
	def, ok := definitions[op]
	if !ok {
		return nil, fmt.Errorf("opcode %d undefined", op)
	}
	return def, nil
}

// Make encodes an instruction from an opcode and operands.
func Make(op Opcode, operands ...int) []byte {
	def, ok := definitions[op]
	if !ok {
		return []byte{}
	}
	instructionLen := 1
	for _, w := range def.OperandWidths {
		instructionLen += w
	}
	instruction := make([]byte, instructionLen)
	instruction[0] = byte(op)
	offset := 1
	for i, o := range operands {
		width := def.OperandWidths[i]
		switch width {
		case 1:
			instruction[offset] = byte(o)
		case 2:
			binary.BigEndian.PutUint16(instruction[offset:], uint16(o))
		}
		offset += width
	}
	return instruction
}

// ReadOperands decodes operands from ins given def.
func ReadOperands(def *Definition, ins Instructions) ([]int, int) {
	operands := make([]int, len(def.OperandWidths))
	offset := 0
	for i, width := range def.OperandWidths {
		switch width {
		case 1:
			operands[i] = int(ReadUint8(ins[offset:]))
		case 2:
			operands[i] = int(ReadUint16(ins[offset:]))
		}
		offset += width
	}
	return operands, offset
}

// ReadUint16 reads a big-endian uint16.
func ReadUint16(ins Instructions) uint16 {
	return binary.BigEndian.Uint16(ins)
}

// ReadUint8 reads a uint8.
func ReadUint8(ins Instructions) uint8 {
	return ins[0]
}

func (ins Instructions) String() string {
	var out string
	i := 0
	for i < len(ins) {
		def, err := Lookup(Opcode(ins[i]))
		if err != nil {
			out += fmt.Sprintf("ERROR: %s\n", err)
			continue
		}
		operands, read := ReadOperands(def, ins[i+1:])
		out += fmt.Sprintf("%04d %s\n", i, fmtInstruction(def, operands))
		i += 1 + read
	}
	return out
}

func fmtInstruction(def *Definition, operands []int) string {
	operandCount := len(def.OperandWidths)
	if len(operands) != operandCount {
		return fmt.Sprintf("ERROR: operand len %d does not match defined %d\n", len(operands), operandCount)
	}
	switch operandCount {
	case 0:
		return def.Name
	case 1:
		return fmt.Sprintf("%s %d", def.Name, operands[0])
	case 2:
		return fmt.Sprintf("%s %d %d", def.Name, operands[0], operands[1])
	}
	return fmt.Sprintf("ERROR: unhandled operandCount for %s\n", def.Name)
}

File: vscode-nexus/src/vm/code.ts
/**
 * Nexus bytecode opcodes and instruction encoding (TypeScript VM).
 */

export type Instructions = Uint8Array;

export enum Opcode {
  OpConstant = 0,
  OpPop,
  OpAdd,
  OpSub,
  OpMul,
  OpDiv,
  OpMod,
  OpTrue,
  OpFalse,
  OpNull,
  OpEqual,
  OpNotEqual,
  OpGreaterThan,
  OpGreaterEqual,
  OpLessThan,
  OpLessEqual,
  OpMinus,
  OpBang,
  OpJump,
  OpJumpNotTruthy,
  OpGetGlobal,
  OpSetGlobal,
  OpGetLocal,
  OpSetLocal,
  OpGetBuiltin,
  OpArray,
  OpHash,
  OpIndex,
  OpCall,
  OpReturnValue,
  OpReturn,
  OpClosure,
  OpGetFree,
  OpCurrentClosure,
  OpSetIndex,
  OpMember,
  OpAnd,
  OpOr,
}

export interface Definition {
  name: string;
  operandWidths: number[];
}

const DEFINITIONS: Record<number, Definition> = {
  [Opcode.OpConstant]: { name: "OpConstant", operandWidths: [2] },
  [Opcode.OpPop]: { name: "OpPop", operandWidths: [] },
  [Opcode.OpAdd]: { name: "OpAdd", operandWidths: [] },
  [Opcode.OpSub]: { name: "OpSub", operandWidths: [] },
  [Opcode.OpMul]: { name: "OpMul", operandWidths: [] },
  [Opcode.OpDiv]: { name: "OpDiv", operandWidths: [] },
  [Opcode.OpMod]: { name: "OpMod", operandWidths: [] },
  [Opcode.OpTrue]: { name: "OpTrue", operandWidths: [] },
  [Opcode.OpFalse]: { name: "OpFalse", operandWidths: [] },
  [Opcode.OpNull]: { name: "OpNull", operandWidths: [] },
  [Opcode.OpEqual]: { name: "OpEqual", operandWidths: [] },
  [Opcode.OpNotEqual]: { name: "OpNotEqual", operandWidths: [] },
  [Opcode.OpGreaterThan]: { name: "OpGreaterThan", operandWidths: [] },
  [Opcode.OpGreaterEqual]: { name: "OpGreaterEqual", operandWidths: [] },
  [Opcode.OpLessThan]: { name: "OpLessThan", operandWidths: [] },
  [Opcode.OpLessEqual]: { name: "OpLessEqual", operandWidths: [] },
  [Opcode.OpMinus]: { name: "OpMinus", operandWidths: [] },
  [Opcode.OpBang]: { name: "OpBang", operandWidths: [] },
  [Opcode.OpJump]: { name: "OpJump", operandWidths: [2] },
  [Opcode.OpJumpNotTruthy]: { name: "OpJumpNotTruthy", operandWidths: [2] },
  [Opcode.OpGetGlobal]: { name: "OpGetGlobal", operandWidths: [2] },
  [Opcode.OpSetGlobal]: { name: "OpSetGlobal", operandWidths: [2] },
  [Opcode.OpGetLocal]: { name: "OpGetLocal", operandWidths: [1] },
  [Opcode.OpSetLocal]: { name: "OpSetLocal", operandWidths: [1] },
  [Opcode.OpGetBuiltin]: { name: "OpGetBuiltin", operandWidths: [1] },
  [Opcode.OpArray]: { name: "OpArray", operandWidths: [2] },
  [Opcode.OpHash]: { name: "OpHash", operandWidths: [2] },
  [Opcode.OpIndex]: { name: "OpIndex", operandWidths: [] },
  [Opcode.OpCall]: { name: "OpCall", operandWidths: [1] },
  [Opcode.OpReturnValue]: { name: "OpReturnValue", operandWidths: [] },
  [Opcode.OpReturn]: { name: "OpReturn", operandWidths: [] },
  [Opcode.OpClosure]: { name: "OpClosure", operandWidths: [2, 1] },
  [Opcode.OpGetFree]: { name: "OpGetFree", operandWidths: [1] },
  [Opcode.OpCurrentClosure]: { name: "OpCurrentClosure", operandWidths: [] },
  [Opcode.OpSetIndex]: { name: "OpSetIndex", operandWidths: [] },
  [Opcode.OpMember]: { name: "OpMember", operandWidths: [2] },
  [Opcode.OpAnd]: { name: "OpAnd", operandWidths: [] },
  [Opcode.OpOr]: { name: "OpOr", operandWidths: [] },
};

export function lookup(op: Opcode): Definition | undefined {
  return DEFINITIONS[op];
}

export function make(op: Opcode, ...operands: number[]): number[] {
  const def = DEFINITIONS[op];
  if (!def) {
    return [];
  }
  let instructionLen = 1;
  for (const w of def.operandWidths) {
    instructionLen += w;
  }
  const instruction = new Array<number>(instructionLen);
  instruction[0] = op;
  let offset = 1;
  for (let i = 0; i < operands.length; i++) {
    const width = def.operandWidths[i]!;
    const o = operands[i]!;
    if (width === 1) {
      instruction[offset] = o & 0xff;
    } else if (width === 2) {
      instruction[offset] = (o >> 8) & 0xff;
      instruction[offset + 1] = o & 0xff;
    }
    offset += width;
  }
  return instruction;
}

export function readUint16(ins: Uint8Array | number[], offset: number): number {
  return ((ins[offset]! & 0xff) << 8) | (ins[offset + 1]! & 0xff);
}

export function readUint8(ins: Uint8Array | number[], offset: number): number {
  return ins[offset]! & 0xff;
}

export function readOperands(
  def: Definition,
  ins: Uint8Array | number[],
  start: number,
): { operands: number[]; bytesRead: number } {
  const operands: number[] = [];
  let offset = 0;
  for (const width of def.operandWidths) {
    if (width === 1) {
      operands.push(readUint8(ins, start + offset));
    } else if (width === 2) {
      operands.push(readUint16(ins, start + offset));
    }
    offset += width;
  }
  return { operands, bytesRead: offset };
}

export function concatInstructions(...parts: number[][]): Uint8Array {
  const total = parts.reduce((n, p) => n + p.length, 0);
  const out = new Uint8Array(total);
  let offset = 0;
  for (const p of parts) {
    out.set(p, offset);
    offset += p.length;
  }
  return out;
}
